chips/stm32f4xx/src/spi.rs

```diff
@@ -0,0 +1,500 @@
+        len: usize,
+    ) -> ReturnCode {
+        if write_buffer.is_none() && read_buffer.is_none() {
+            return ReturnCode::EINVAL;
+        }
+
+        self.hold_low();
+
+        let mut count: usize = len;
+        write_buffer
+            .as_ref()
+            .map(|buf| count = cmp::min(count, buf.len()));
+        read_buffer
+            .as_ref()
+            .map(|buf| count = cmp::min(count, buf.len()));
+
+        self.dma_len.set(count);
+
+        self.transfers_in_progress.set(0);
+
+        read_buffer.map(|rx_buffer| {
+            self.transfers_in_progress
+                .set(self.transfers_in_progress.get() + 1);
+            self.rx_dma.map(move |dma| {
+                dma.do_transfer(rx_buffer, count);
+            });
+            self.enable_rx();
+        });
+
+        write_buffer.map(|tx_buffer| {
+            self.transfers_in_progress
+                .set(self.transfers_in_progress.get() + 1);
+            self.tx_dma.map(move |dma| {
+                dma.do_transfer(tx_buffer, count);
+            });
+            self.enable_tx();
+        });
+
+        ReturnCode::SUCCESS
+    }
+}
+
+impl spi::SpiMaster for Spi<'a> {
+    type ChipSelect = PinId;
+
+    fn set_client(&self, client: &'static SpiMasterClient) {
+        self.master_client.set(client);
+    }
+
+    fn init(&self) {
+        // // enable error interrupt (used only for debugging)
+        // self.registers.cr2.modify(CR2::ERRIE::SET);
+
+        self.registers.cr1.modify(
+            // 2 line unidirectional mode
+            CR1::BIDIMODE::CLEAR +
+            // Select as master
+            CR1::MSTR::SET +
+            // Software slave management
+            CR1::SSM::SET +
+            CR1::SSI::SET +
+            // 8 bit data frame format
+            CR1::DFF::CLEAR +
+            // Enable
+            CR1::SPE::SET,
+        );
+    }
+
+    fn is_busy(&self) -> bool {
+        self.registers.sr.is_set(SR::BSY)
+    }
+
+    fn write_byte(&self, out_byte: u8) {
+        // loop till TXE (Transmit Buffer Empty) becomes 1
+        while !self.registers.sr.is_set(SR::TXE) {}
+
+        self.registers.dr.modify(DR::DR.val(out_byte as u32));
+    }
+
+    fn read_byte(&self) -> u8 {
+        self.read_write_byte(0)
+    }
+
+    fn read_write_byte(&self, val: u8) -> u8 {
+        self.write_byte(val);
+
+        // loop till RXNE becomes 1
+        while !self.registers.sr.is_set(SR::RXNE) {}
+
+        self.registers.dr.read(DR::DR) as u8
+    }
+
+    fn read_write_bytes(
+        &self,
+        write_buffer: &'static mut [u8],
+        read_buffer: Option<&'static mut [u8]>,
+        len: usize,
+    ) -> ReturnCode {
+        // If busy, don't start
+        if self.is_busy() {
+            return ReturnCode::EBUSY;
+        }
+
+        self.read_write_bytes(Some(write_buffer), read_buffer, len)
+    }
+
+    // We *only* support 1Mhz
+    fn set_rate(&self, rate: u32) -> u32 {
```

Comment:
Please add docs that the function panics if rate != 1_000_000

Reply:
Done.